File: objc/Tests/TestM76JcampConformance.m
#import <Foundation/Foundation.h>
#import "Testing.h"
#import "Core/TTIOSignalArray.h"
#import "Spectra/TTIOUVVisSpectrum.h"
#import "ValueClasses/TTIOEncodingSpec.h"
#import "Export/TTIOJcampDxWriter.h"
#import "Export/TTIOJcampDxEncoding.h"
#import <unistd.h>

// M76 byte-parity conformance test for the ObjC JCAMP-DX compressed
// writer. Each mode (PAC / SQZ / DIF) has a matching golden fixture
// under conformance/jcamp_dx/. Python and Java ship the analogous
// tests — together they form the M76 cross-language byte-parity gate.

static NSString *conformanceDir(void)
{
    // objc/Tests runs from the build/test dir. Walk upward looking
    // for conformance/jcamp_dx so the test works from any sane CWD.
    NSFileManager *fm = [NSFileManager defaultManager];
    NSString *here = [fm currentDirectoryPath];
    for (int up = 0; up < 6; up++) {
        NSString *candidate = [[here
                stringByAppendingPathComponent:@"conformance"]
                stringByAppendingPathComponent:@"jcamp_dx"];
        BOOL isDir = NO;
        if ([fm fileExistsAtPath:candidate isDirectory:&isDir] && isDir) {
            return candidate;
        }
        here = [here stringByDeletingLastPathComponent];
        if ([here isEqualToString:@"/"] || here.length == 0) break;
    }
    return nil;
}

static TTIOSignalArray *float64Arr(const double *src, NSUInteger n)
{
    NSData *buf = [NSData dataWithBytes:src length:n * sizeof(double)];
    TTIOEncodingSpec *enc =
        [TTIOEncodingSpec specWithPrecision:TTIOPrecisionFloat64
                       compressionAlgorithm:TTIOCompressionZlib
                                  byteOrder:TTIOByteOrderLittleEndian];
    return [[TTIOSignalArray alloc] initWithBuffer:buf
                                            length:n
                                          encoding:enc
                                              axis:nil];
}

static TTIOUVVisSpectrum *ramp25Fixture(void)
{
    const NSUInteger N = 25;
    double *wl     = malloc(N * sizeof(double));
    double *absorb = malloc(N * sizeof(double));
    for (NSUInteger i = 0; i < N; i++) {
        wl[i]     = 200.0 + (double)i * 10.0;
        double a  = (double)i;
        double b  = 24.0 - (double)i;
        absorb[i] = (a < b) ? a : b;
    }
    TTIOSignalArray *wlA = float64Arr(wl, N);
    TTIOSignalArray *abA = float64Arr(absorb, N);
    free(wl); free(absorb);
    NSError *initErr = nil;
    TTIOUVVisSpectrum *spec = [[TTIOUVVisSpectrum alloc]
            initWithWavelengthArray:wlA
                    absorbanceArray:abA
                       pathLengthCm:1.0
                            solvent:@"water"
                      indexPosition:0
                    scanTimeSeconds:0.0
                              error:&initErr];
    if (!spec) {
        NSLog(@"UVVisSpectrum init failed: %@", initErr);
    }
    return spec;
}

static void checkOneMode(TTIOJcampDxEncoding mode,
                          NSString *modeName,
                          NSString *fixtureFile,
                          NSString *confDir)
{
    NSString *golden = [confDir stringByAppendingPathComponent:fixtureFile];
    if (![[NSFileManager defaultManager] fileExistsAtPath:golden]) {
        // Golden missing — treat as skip (Python + Java gate is the
        // first line of defense, we don't want the ObjC suite to fail
        // hard when fixtures are absent from a partial checkout).
        pass(YES, "skip: fixture missing %s", [golden UTF8String]);
        return;
    }
    NSString *outPath = [NSString stringWithFormat:@"/tmp/ttio_m76_%d_%@.jdx",
                         (int)getpid(), modeName];

    NSError *err = nil;
    BOOL ok = [TTIOJcampDxWriter writeUVVisSpectrum:ramp25Fixture()
                                              toPath:outPath
                                               title:@"m76 ramp-25"
                                            encoding:mode
                                               error:&err];
    pass(ok, "write %s: %s", [modeName UTF8String],
         err ? [[err localizedDescription] UTF8String] : "ok");

    NSData *produced = [NSData dataWithContentsOfFile:outPath];
    NSData *expected = [NSData dataWithContentsOfFile:golden];
    BOOL match = [produced isEqualToData:expected];
    if (!match) {
        NSString *producedStr = [[NSString alloc] initWithData:produced
                                                      encoding:NSUTF8StringEncoding];
        NSString *expectedStr = [[NSString alloc] initWithData:expected
                                                      encoding:NSUTF8StringEncoding];
        NSLog(@"byte-parity drift on %@ encoder.\n--- expected ---\n%@--- produced ---\n%@",
              modeName, expectedStr, producedStr);
    }
    pass(match, "byte-parity %s", [modeName UTF8String]);

    unlink([outPath fileSystemRepresentation]);
}

void testM76JcampConformance(void)
{
    NSString *confDir = conformanceDir();
    if (!confDir) {
        pass(YES, "skip: conformance/jcamp_dx not reachable from CWD");
        return;
    }
    checkOneMode(TTIOJcampDxEncodingPAC, @"pac", @"uvvis_ramp25_pac.jdx", confDir);
    checkOneMode(TTIOJcampDxEncodingSQZ, @"sqz", @"uvvis_ramp25_sqz.jdx", confDir);
    checkOneMode(TTIOJcampDxEncodingDIF, @"dif", @"uvvis_ramp25_dif.jdx", confDir);
}

File: python/tests/test_m76_jcamp_conformance.py
"""M76 — byte-parity conformance check for the JCAMP-DX compressed writer.

Each mode (PAC / SQZ / DIF) has a matching golden fixture under
``conformance/jcamp_dx/``. This test regenerates the same output
in-process and asserts byte-for-byte equality. Java and Objective-C
ship the analogous tests — together they form the M76 cross-language
byte-parity gate.

If this test starts failing, either:

1. The Python encoder changed (intentional) — regenerate the
   goldens via ``python conformance/jcamp_dx/generate.py`` **and**
   push through the Java + ObjC writers so all three re-align.
2. The Python encoder changed (unintentional) — revert.

Silently-updating the goldens defeats the whole purpose of the gate.
"""
from __future__ import annotations

from pathlib import Path

import numpy as np
import pytest

from ttio import AxisDescriptor, SignalArray, UVVisSpectrum
from ttio.exporters.jcamp_dx import write_uv_vis_spectrum


_CONFORMANCE = Path(__file__).resolve().parents[2] / "conformance" / "jcamp_dx"


def _ramp25_fixture() -> UVVisSpectrum:
    wl = np.linspace(200.0, 440.0, 25, dtype=np.float64)
    absorb = np.array([min(i, 24 - i) for i in range(25)], dtype=np.float64)
    return UVVisSpectrum(
        signal_arrays={
            UVVisSpectrum.WAVELENGTH: SignalArray.from_numpy(
                wl, axis=AxisDescriptor("wavelength", "nm"),
            ),
            UVVisSpectrum.ABSORBANCE: SignalArray.from_numpy(
                absorb, axis=AxisDescriptor("absorbance", ""),
            ),
        },
        path_length_cm=1.0,
        solvent="water",
    )


@pytest.mark.parametrize("mode", ["pac", "sqz", "dif"])
def test_python_writer_matches_golden(tmp_path: Path, mode: str) -> None:
    golden = _CONFORMANCE / f"uvvis_ramp25_{mode}.jdx"
    if not golden.is_file():
        pytest.skip(f"golden fixture missing: {golden}")
    out = tmp_path / f"gen_{mode}.jdx"
    write_uv_vis_spectrum(
        _ramp25_fixture(), out, title="m76 ramp-25", encoding=mode,
    )
    produced = out.read_bytes()
    expected = golden.read_bytes()
    assert produced == expected, (
        f"byte-parity drift on {mode.upper()} encoder.\n"
        f"--- expected ({golden.name}) ---\n{expected.decode(errors='replace')}"
        f"--- produced ---\n{produced.decode(errors='replace')}"
    )
